Update _store_based_barrier implementation to reduce load on rank 0

H-Huang opened the pull request:
Summary: 

Update from using add() which makes rank 0 overloaded with requests to a single request every 10 seconds to handle the last joined worker
Added optional logging_interval arg to _store_based_barrier

Test Plan:
```
pytest test/distributed/test_c10d_common.py -vsk test_store_based_barrier
```

Reviewed By: rohan-varma

Differential Revision: D44430531

kumpera commented on torch/distributed/distributed_c10d.py:
```diff
@@ -474,44 +474,46 @@ def _store_based_barrier(rank, store, timeout):
-    store_key = "{}:{}".format(STORE_BASED_BARRIER_PREFIX, _world.group_count)
-    store.add(store_key, 1)
-    logger.info("Added key: {} to store for rank: {}".format(store_key, rank))
+    global_key = "{}:{}".format(STORE_BASED_BARRIER_PREFIX, _world.group_count)
```
We could use f strings here.

H-Huang replied:
done, thanks!

kumpera commented on torch/distributed/distributed_c10d.py:
```diff
@@ -474,44 +474,46 @@ def _store_based_barrier(rank, store, timeout):
+    start = time.time()
+    while True:
+        try:
+            worker_count = store.add(global_key, 0)
```
We could move this to the exception/log path

H-Huang replied:
done

kumpera commented on torch/distributed/distributed_c10d.py:
```diff
@@ -474,44 +474,46 @@ def _store_based_barrier(rank, store, timeout):
+            worker_count = store.add(global_key, 0)
+            # This will throw an exception after 10 seconds in which we print out
+            # the status of the group or time out officially, throwing runtime error
+            store.wait(all_keys, timedelta(seconds=10))
```
This should be a global prop in c10d so we can patch it for testing purposes.

And it should be `min(10, timeout)` otherwise asking for a 1 seconds timeout will force a minimum of 10 seconds.

H-Huang replied:
to my knowledge, we don't currently have a way to validate that a log was outputted. So I am still not sure how this would be tested. We can validate the timeout works correctly by writing tests for `wait()`(i've only just done it manually right now) instead of `_store_based_barrier()`

kumpera commented on test/distributed/test_store.py:
```diff
@@ -311,6 +315,28 @@ def test_multi_worker_with_fixed_world_size(self):
+class StoreBasedBarrierTest(MultiThreadedTestCase):
+    @property
+    def world_size(self):
+        return 1
```
We're not really testing much here with a world_size of one :)

H-Huang replied:
Yeah, the only thing is that the vast majority of our tests use `FileStore`. So this would also be directly testing that `store_based_barrier` works with `HashStore` and `TcpStore`.

I was not able to make it larger than 1 for TCPStore when it was using `MultiThreadedTestCase`, so I need to update it. I was planning on splitting out `init_process_group` from that class

kumpera commented on test/distributed/test_store.py:
```diff
@@ -311,6 +315,28 @@ def test_multi_worker_with_fixed_world_size(self):
+        self._spawn_threads()
+
+    def test_hash_store_based_barrier(self):
+        _store_based_barrier(dist.get_rank(), self.hash_store, timedelta(seconds=20))
```
We need a test for the timeout case.

Plus we need to lower the minimum way from 10 seconds in c10d to something smaller like 1s to test the looping behavior more aggressively.

H-Huang committed:
Update _store_based_barrier implementation to reduce load on rank 0 (#98000)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/98000

Update from using `add()` which makes rank 0 overloaded with requests to a single request every 10 seconds to handle the last joined worker

Added optional `logging_interval` arg to `_store_based_barrier`

Test Plan:
Added `test_store_based_barrier` test case

```
buck build //caffe2/test/distributed:c10d --show-full-output
<SHOW_FULL_OUTPUT.par> -r test_store_based_barrier
```

Reviewed By: rohan-varma

Differential Revision: D44430531

fbshipit-source-id: 366c95269328050d813204447cb028f6996d77d0

kumpera reviewed (approved):
Looks great!

H-Huang commented:
@pytorchbot merge

H-Huang commented:
@pytorch merge -f "mac failures are unrelated: 'ModuleNotFoundError: no module named sympy'"
